contracts/upgradeable_contracts/arbitrary_message/MessageProcessor.sol

```diff
@@ -0,0 +1,78 @@
+pragma solidity 0.4.24;
+
+import "./BalanceHandler.sol";
+
+contract MessageProcessor is BalanceHandler {
+    uint256 internal constant PASS_MESSAGE_GAS = 100000;
+    bytes32 internal constant MESSAGE_SENDER = keccak256(abi.encodePacked("messageSender"));
+
+    function messageSender() external view returns (address) {
+        return addressStorage[MESSAGE_SENDER];
+    }
+
+    function setMessageSender(address _sender) internal {
+        addressStorage[MESSAGE_SENDER] = _sender;
+    }
+
+    function processMessage(
+        address sender,
+        address executor,
+        bytes32 txHash,
+        uint256 gasLimit,
+        bytes1 dataType,
+        uint256 gasPrice,
+        bytes memory data
+    ) internal {
+        bool status;
+
+        if (dataType == 0x00) {
+            require(isMessageProcessorSubsidizedMode());
+            status = _passMessage(sender, executor, data, gasLimit);
+        } else if (dataType == 0x01) {
+            require(!isMessageProcessorSubsidizedMode());
+            require(gasPrice == tx.gasprice);
+            status = _defrayAndPassMessage(sender, executor, data, gasLimit);
+        } else if (dataType == 0x02) {
+            require(!isMessageProcessorSubsidizedMode());
+            status = _defrayAndPassMessage(sender, executor, data, gasLimit);
+        } else {
+            status = false;
+        }
+
+        emitEventOnMessageProcessed(sender, executor, txHash, status);
+    }
+
+    function _defrayAndPassMessage(address _sender, address _contract, bytes _data, uint256 _gas)
+        internal
+        returns (bool)
+    {
+        uint256 fee = PASS_MESSAGE_GAS.add(_gas).mul(tx.gasprice);
+        require(balanceOf(_sender) >= fee);
+        require(address(this).balance >= fee);
+
+        setBalanceOf(_sender, balanceOf(_sender).sub(fee));
+
+        bool status = _passMessage(_sender, _contract, _data, _gas);
+
+        msg.sender.transfer(fee);
+
+        return status;
+    }
+
+    function _passMessage(address _sender, address _contract, bytes _data, uint256 _gas) internal returns (bool) {
+        if (_contract == address(this)) {
+            //Special check to handle invocation of withdrawFromDeposit
+            if (isWithdrawFromDepositSelector(_data)) {
+                setAccountForAction(_sender);
+            }
+        }
+        setMessageSender(_sender);
+        bool status = _contract.call.gas(_gas)(_data);
```

Comment:
Is it safe if we don't handle the case when the method returns some data after invocation?

Reply:
I'm not sure I fully understood the question, `contract.call` will return `true` if execution is correct and `false` if it is reverted and we show this result in the emitted event later. If we want to get the returned data from the method we should use the assembly call, but I'm not sure why we need it

Reply:
If you look at the `call` function [description](https://solidity.readthedocs.io/en/v0.5.11/units-and-global-variables.html#members-of-address-types) you will see that it returns `(bool, bytes memory)`. But in the discussed code we are expecting only a bool value. So, the question is if the calling method returns something will this be skipped successfully even if we don't allocate memory for this data explicitly.

Reply:
Oh I see. Actually, what you mention about `call` function is valid for solidity version > `0.5.0`. For [version `0.4.24`](https://solidity.readthedocs.io/en/v0.4.24/units-and-global-variables.html#address-related) `call` function only returns `(bool)`, so I think we don't need to worry about a possible returned data

Reply:
Thanks! Sorry for my inattention.